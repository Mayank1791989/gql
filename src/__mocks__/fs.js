/* @flow */
import MemoryFileSystem from 'memory-fs';
import path from 'path';

function createNewMemoryFileSystem(): MemoryFileSystem {
  const _fs = new MemoryFileSystem();
  // Note: binding all methods of _fs
  // some libs use individual fs method
  // e.g const writeFileSync = fs.writeFileSync
  //  writeFileSync() // this will call method with wrong this
  // happening in find-config package
  Object.keys(MemoryFileSystem.prototype).forEach((key) => {
    if (typeof _fs[key] === 'function') {
      _fs[key] = _fs[key].bind(_fs);
    }
  });
  return _fs;
}

const fs = createNewMemoryFileSystem();

// This is a custom function that our tests can use during setup to create
// files in memory
function __setMockFiles(newMockFiles) {
  Object.keys(newMockFiles).forEach((file) => {
    const dirname = path.dirname(file);
    fs.mkdirpSync(dirname);
    fs.writeFileSync(file, newMockFiles[file], 'utf8');
  });
}

fs.__setMockFiles = __setMockFiles;

module.exports = fs;
