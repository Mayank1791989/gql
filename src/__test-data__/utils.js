/* @flow */
import path from 'path';
import { dedent } from 'dentist';
import splitLines from '../shared/splitLines';

export function getSchema() {
  jest.mock('../shared/watch');
  const watch = require('../shared/watch').default;
  const GQLService = require('../').GQLService;

  const gql = new GQLService({ cwd: path.resolve('src/__test-data__/') });
  // $FlowDisableNextLine
  watch.__triggerChange();
  return gql._schemaBuilder.getSchema();
}

export function getDefLocations() {
  return {
    Player: {
      start: { line: 12, column: 1 },
      end: { line: 16, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_id: {
      start: { line: 13, column: 3 },
      end: { line: 13, column: 10 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_image_arg_size: {
      start: { line: 15, column: 9 },
      end: { line: 15, column: 19 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Team: {
      start: { line: 17, column: 1 },
      end: { line: 20, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Node: {
      start: { line: 8, column: 1 },
      end: { line: 10, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Edge: {
      start: { line: 27, column: 1 },
      end: { line: 30, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Mutation: {
      start: { line: 2, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreate: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 63 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput: {
      start: { line: 6, column: 1 },
      end: { line: 9, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput_id: {
      start: { line: 8, column: 3 },
      end: { line: 8, column: 10 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },

    Query: {
      start: { line: 2, column: 1 },
      end: { line: 6, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Query_viewer: {
      start: { line: 5, column: 3 },
      end: { line: 5, column: 18 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Role: {
      start: { line: 1, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/enums.gql'),
    },

    CustomScalar: {
      start: { line: 1, column: 1 },
      end: { line: 1, column: 20 },
      path: path.resolve('src/__test-data__/schema/scalars.gql'),
    },

    customDirective: {
      start: { line: 2, column: 1 },
      end: { line: 7, column: 25 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
    customDirective_argIf: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 15 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
  };
}

export function getRefLocations() {
  return {
    Player: [
      {
        start: { line: 14, column: 11 },
        end: { line: 14, column: 17 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 24, column: 7 },
        end: { line: 24, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 12, column: 6 },
        end: { line: 12, column: 12 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Edge: [
      {
        start: { line: 28, column: 11 },
        end: { line: 28, column: 15 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Role: [
      {
        start: { line: 10, column: 9 },
        end: { line: 10, column: 13 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 34, column: 9 },
        end: { line: 34, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 36, column: 15 },
        end: { line: 36, column: 19 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 6 },
        end: { line: 1, column: 10 },
        path: path.resolve('src/__test-data__/schema/enums.gql'),
      },
    ],

    CustomScalar: [
      {
        start: { line: 35, column: 10 },
        end: { line: 35, column: 22 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 8 },
        end: { line: 1, column: 20 },
        path: path.resolve('src/__test-data__/schema/scalars.gql'),
      },
    ],
  };
}

export function getHints() {
  return {
    ObjectTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    OutputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    CompositeTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    InputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: '',
        text: 'PlayerCreateInput',
        type: 'Input',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    InterfaceTypes: [
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
    ],

    PlayerTypeFields: [
      {
        description: '',
        text: 'id',
        type: 'ID!',
      },
      {
        description: '',
        text: 'name',
        type: 'String!',
      },
      {
        description: '',
        text: 'image',
        type: 'String!',
      },
    ],

    PlayerTypeImageFieldArgs: [
      {
        description: '',
        text: 'size',
        type: 'Int!',
      },
    ],

    TypesImplementsNode: [
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
    ],

    PossibleMutations: [
      {
        description: 'Player Create contains all allowed mutations',
        text: 'PlayerCreate',
        type: 'PlayerCreatePayload',
      },
    ],

    QueryFields: [
      {
        description: '',
        text: 'node',
        type: 'Node',
      },
      {
        description: '',
        text: 'nodes',
        type: '[Node]',
      },
      {
        description: '',
        text: 'viewer',
        type: 'Viewer!',
      },
      {
        description: 'Access the current type schema of this server.',
        text: '__schema',
        type: '__Schema!',
      },
      {
        description: 'Request the type information of a single type.',
        text: '__type',
        type: '__Type',
      },
    ],

    EnumRoleValues: [
      {
        description: '',
        text: 'roleA',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleB',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleC',
        type: 'Role',
      },
    ],
  };
}

export function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
