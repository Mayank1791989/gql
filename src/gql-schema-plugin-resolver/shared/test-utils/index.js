/* @flow */
export { default as mockTestResolverParser } from './mockTestResolverParser';
