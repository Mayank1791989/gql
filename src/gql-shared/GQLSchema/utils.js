/* @flow */
import {
  type TypeNode,
  type NamedTypeNode,
  type EnumValueDefinitionNode,
  type FieldDefinitionNode,
  getDirectiveValues,
  GraphQLDeprecatedDirective,
  Kind,
} from 'graphql';

export function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (
    namedType.kind === Kind.LIST_TYPE ||
    namedType.kind === Kind.NON_NULL_TYPE
  ) {
    namedType = namedType.type;
  }
  return namedType;
}

/**
 * Given a field or enum value node, returns the string value for the
 * deprecation reason.
 */
export function getDeprecationReason(
  node: EnumValueDefinitionNode | FieldDefinitionNode,
): ?string {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node);
  return deprecated && (deprecated.reason: any);
}
