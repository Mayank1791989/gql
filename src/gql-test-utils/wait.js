/* @flow */
export async function waitFor(
  condition: () => boolean,
  timeout?: number = 5000,
  interval?: number = 100,
): Promise<void> {
  // start timeout
  const id = setTimeout(() => {
    throw new Error('waitFor fn timed out');
  }, timeout);

  // wait for condition
  while (!condition()) {
    // eslint-disable-next-line no-await-in-loop
    await wait(interval);
  }

  // clear
  clearTimeout(id);
}

export const conditions = {
  mockFnCall: (mockFn: JestMockFn<any, any>) => {
    const currCalls = mockFn.mock.calls.length;
    return () => {
      return mockFn.mock.calls.length > currCalls;
    };
  },
};

export function wait(timeInMsec: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, timeInMsec);
  });
}
