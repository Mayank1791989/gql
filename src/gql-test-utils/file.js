/* @flow */
import os from 'os';
import fs from 'fs-extra';
import path from 'path';

export const TEST_TEMP_DIR = path.join(
  // NOTE: os.tmpdir doesnt return realpath
  // so causing issues on osx ci server
  fs.realpathSync(os.tmpdir()),
  'gql-test',
);

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, `test-${Date.now()}`);

  // @TOOD: should we emptyDirSync?
  // NOTE: ensure directory is empty
  // when tests are run in watch mode we want to make sure files not present from last run
  // fs.emptyDirSync(dir);

  Object.keys(files).forEach(file => {
    const content = files[file];
    fs.outputFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  fs.removeSync(TEST_TEMP_DIR);
}

export function replaceTempDir(filepath: string, str: string): string {
  return path.relative(TEST_TEMP_DIR, filepath).replace(/test-\d+/u, str);
}
