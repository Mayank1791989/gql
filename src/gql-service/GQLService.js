/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log, { type LogListener } from 'gql-shared/log';
import { Source } from 'graphql';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

export type { Options as GQLServiceOptions };

export default class GQLService extends GQLBaseService {
  _config: GQLConfig;
  _watcher: GQLWatcher;

  _services: [GQLSchemaService, GQLQueryService] = [];

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    const schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    schemaService.onChange(this._triggerChange);
    schemaService.onError(this._triggerError);
    this._services[0] = schemaService;

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._services[1] = queryService;
    }
  }

  onLog(listener: LogListener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await Promise.all(this._services.map(service => service.start()));
  }

  async _handleStop() {
    await Promise.all(this._services.map(service => service.stop()));
  }

  getSchema(): GQLSchema {
    const [schemaService] = this._services;
    return schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      return this._services.reduce((acc, service) => {
        return acc.concat(service.getErrors());
      }, []);
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): $ReadOnlyArray<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('autocomplete response');
      const position = {
        line: params.position.line,
        column: params.position.column - 1,
      };
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getHintsAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('autocomplete response');

      return results;
    }, []);
  }

  getDef(params: CommandParams): $ReadOnlyArray<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('getDef request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('getDef response');
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getDefinitionAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('getDef response');

      return results;
    }, []);
  }

  findRefs(params: CommandParams): $ReadOnlyArray<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('findRefs response');
      const results = this._services.reduce((acc, service) => {
        acc.push(
          ...service.findRefsOfTokenAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('findRefs response');

      return results;
    }, []);
  }

  getInfo(params: CommandParams): $ReadOnlyArray<GQLInfo> {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      if (!this._isRunning) {
        return [];
      }

      logger.time('getInfo response');
      const result = this._services.reduce((acc, service) => {
        acc.push(
          ...service.getInfoOfTokenAtPosition({
            source: new Source(params.sourceText, params.sourcePath),
            position: params.position,
          }),
        );
        return acc;
      }, []);
      logger.timeEnd('getInfo response');

      return result;
    }, []);
  }
}
