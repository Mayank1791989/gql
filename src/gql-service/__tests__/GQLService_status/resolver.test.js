/* @flow */
import GQLService from '../../GQLService';
import { mockTestResolverParser } from 'gql-schema-plugin-resolver/shared/test-utils';
import { status as _status } from './utils';
import { createTempFiles } from 'gql-test-utils/file';
import { waitFor, conditions } from 'gql-test-utils/wait';
import path from 'path';
import fs from 'fs-extra';

beforeEach(() => {
  mockTestResolverParser('gql-resolver-parser-test');
});

const schemaFiles = {
  'schema/schema.gql': `
    type Query {
      viewer: Viewer
    }
    type Viewer {
      name: String
    }
  `,
};

describe('report error when schema changed', () => {
  async function createService() {
    const rootDir = createTempFiles({
      '.git/tmp': 'to trigger watchman events',
      '.gqlconfig': JSON.stringify({
        schema: {
          files: ['schema/*.gql'],
          validate: {
            config: {
              NoUnusedTypeDefinition: 'off',
            },
          },
          plugins: [
            [
              'resolver',
              {
                files: 'resolver/*.js',
                parser: 'test',
              },
            ],
          ],
        },
      }),
      ...schemaFiles,
    });

    const gql = new GQLService({
      configDir: rootDir,
      watch: true,
    });
    gql.onError(err => {
      throw err;
    });

    await gql.start();
    return { gql, rootDir };
  }

  test('schema', async () => {
    const { gql, rootDir } = await createService();
    const onChange = jest.fn();
    gql.onChange(onChange);

    fs.outputFile(
      path.join(rootDir, 'schema/scalar.gql'),
      'scalar customScalar',
    );
    await waitFor(conditions.mockFnCall(onChange));

    expect(gql.status()).toMatchSnapshot();

    fs.outputFile(
      path.join(rootDir, 'resolver/test.js'),
      'scalars[customScalar] = func;',
    );
    await waitFor(conditions.mockFnCall(onChange));

    expect(gql.status()).toEqual([]);

    await gql.stop();
  });

  // test('global', () => {});

  // test('schema', () => {});
});

describe('UniqueResolvers', () => {
  test('Single file', async () => {
    const errors = await status({
      ...schemaFiles,
      'resolvers/test.js': `
        resolvers[Test/id] = func;
        resolvers[Test/id] = func;
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('Multiple files', async () => {
    expect(
      await status({
        ...schemaFiles,
        'resolvers/testA.js': `
          resolvers[Test/id] = func;
        `,
        'resolvers/testB.js': `
          resolvers[Test/id] = func;
        `,
      }),
    ).toMatchSnapshot();
  });

  test('Can be disabled for some types', async () => {
    expect(
      await status(
        {
          ...schemaFiles,
          'resolvers/testA.js': `
            resolvers[Test/id] = func;
          `,
          'resolvers/testB.js': `
            resolvers[Test/id] = func;
          `,
        },
        {
          validate: {
            config: {
              UniqueResolvers: [
                'error',
                {
                  ignore: [{ typeName: 'Test' }],
                },
              ],
            },
          },
        },
      ),
    ).toMatchSnapshot();
  });

  test('Can be disabled for some fields', async () => {
    expect(
      await status(
        {
          ...schemaFiles,
          'resolvers/testA.js': `
            resolvers[Test/id] = func;
            resolvers[Test/name] = func;
          `,
          'resolvers/testB.js': `
            resolvers[Test/id] = func;
            resolvers[Test/name] = func;
          `,
        },
        {
          validate: {
            config: {
              UniqueResolvers: [
                'error',
                {
                  ignore: [{ typeName: 'Test', fieldName: 'name' }],
                },
              ],
            },
          },
        },
      ),
    ).toMatchSnapshot();
  });

  test.skip('Throw if invalid options passed', async () => {
    try {
      await status(
        {
          ...schemaFiles,
          'resolvers/testA.js': `
            scalars[Test/id] = func;
          `,
        },
        {
          validate: {
            config: {
              UniqueResolvers: ['error', 'tester'],
            },
          },
        },
      );
    } catch (err) {
      console.log(err);
    }
  });
});

describe('PossibleResolvers', () => {});

describe('ProvidedRequiredResolvers', () => {
  test('Scalars', async () => {
    const errors = await status({
      ...schemaFiles,
      'schema/scalars.gql': `
        scalar customScalarOne
        scalar customScalarTwo
      `,
      'resolvers/scalars.js': `
        scalars[customScalarOne] = func;
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('Directives', async () => {
    const errors = await status({
      ...schemaFiles,
      'schema/directives.gql': `
        directive @customDirective(if: Boolean) on FIELD
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('Mutations', async () => {
    const errors = await status({
      ...schemaFiles,
      'schema/scalars.gql': `
        type Mutation {
          name(input: Test): String!
          TestCreate(input: TestCreateInput): TestCreatePayload
        }

        input TestCreateInput {
          name: String
        }

        type TestCreatePayload {
          name: String
        }
      `,
      'resolvers/mutations.js': `
        resolvers[Mutation/name] = func;
      `,
    });
    expect(errors).toMatchSnapshot();
  });

  test('Mutation extensions', async () => {
    const errors = await status({
      ...schemaFiles,
      'schema/scalars.gql': `
        type Mutation {
          name(value: String): String
        }

        extend type Mutation {
          TestCreate(input: TestCreateInput): TestCreatePayload
        }

        input TestCreateInput {
          name: String
        }

        type TestCreatePayload {
          name: String
        }
      `,
      'resolvers/mutations.js': `
        resolvers[Mutation/name] = func;
      `,
    });
    expect(errors).toMatchSnapshot();
  });
});

function status(files, params) {
  return _status(files, {
    resolverConfig: {
      files: 'resolvers/**/*.js',
      parser: 'test',
      ...params,
    },
  });
}
