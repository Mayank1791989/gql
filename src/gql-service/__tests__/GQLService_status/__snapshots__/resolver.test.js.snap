// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProvidedRequiredResolvers Directives 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 12,
        "line": 1,
        "path": "$ROOT_DIR/schema/directives.gql",
      },
    ],
    "message": "Missing directive resolver (ProvidedRequiredResolvers)",
    "severity": "error",
  },
]
`;

exports[`ProvidedRequiredResolvers Mutation extensions 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 6,
        "path": "$ROOT_DIR/schema/scalars.gql",
      },
    ],
    "message": "Missing field resolver (ProvidedRequiredResolvers)",
    "severity": "error",
  },
]
`;

exports[`ProvidedRequiredResolvers Mutations 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 3,
        "path": "$ROOT_DIR/schema/scalars.gql",
      },
    ],
    "message": "Missing field resolver (ProvidedRequiredResolvers)",
    "severity": "error",
  },
  Object {
    "locations": Array [
      Object {
        "column": 15,
        "line": 2,
        "path": "$ROOT_DIR/schema/scalars.gql",
      },
    ],
    "message": "Type \\"Test\\" not found.",
    "severity": "error",
  },
]
`;

exports[`ProvidedRequiredResolvers Scalars 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 8,
        "line": 2,
        "path": "$ROOT_DIR/schema/scalars.gql",
      },
    ],
    "message": "Missing scalar resolver (ProvidedRequiredResolvers)",
    "severity": "error",
  },
]
`;

exports[`UniqueResolvers Can be disabled for some fields 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/resolvers/testA.js",
      },
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/resolvers/testB.js",
      },
    ],
    "message": "Multiple resolvers not allowed. (UniqueResolvers)",
    "severity": "error",
  },
]
`;

exports[`UniqueResolvers Can be disabled for some types 1`] = `Array []`;

exports[`UniqueResolvers Multiple files 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/resolvers/testA.js",
      },
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/resolvers/testB.js",
      },
    ],
    "message": "Multiple resolvers not allowed. (UniqueResolvers)",
    "severity": "error",
  },
]
`;

exports[`UniqueResolvers Single file 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 1,
        "line": 1,
        "path": "$ROOT_DIR/resolvers/test.js",
      },
      Object {
        "column": 1,
        "line": 2,
        "path": "$ROOT_DIR/resolvers/test.js",
      },
    ],
    "message": "Multiple resolvers not allowed. (UniqueResolvers)",
    "severity": "error",
  },
]
`;

exports[`report error when schema changed schema 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 8,
        "line": 1,
        "path": "$ROOT_DIR/schema/scalar.gql",
      },
    ],
    "message": "Missing scalar resolver (ProvidedRequiredResolvers)",
    "severity": "error",
  },
]
`;
